internal_ws/yktrace/src/swt.rs

```diff
@@ -1,45 +1,75 @@
 //! Software tracing via ykrustc.
 
 use super::{SirTrace, ThreadTracer, ThreadTracerImpl};
-use crate::{errors::InvalidTraceError, sir::SIR, SirLoc};
+use crate::{errors::InvalidTraceError, sir::SirLoc};
 use libc;
 use std::convert::TryFrom;
-use ykpack::Local;
 
 /// Softare thread tracer.
 struct SWTThreadTracer;
 
+/// Stop tracing on the current thread.
 impl ThreadTracerImpl for SWTThreadTracer {
     fn stop_tracing(&mut self) -> Result<SirTrace, InvalidTraceError> {
-        let mut len = 0;
-        let buf = unsafe { yk_swt_stop_tracing_impl(&mut len) };
-        if buf as usize == 0 {
-            Err(InvalidTraceError::InternalError)
-        } else {
-            // When we make a SirTrace, we convert all of the locations from SwtLoc to SirLoc.
-            let locs = (0..len)
-                .map(|idx| {
-                    let idx = isize::try_from(idx).unwrap();
-                    let swt_loc = unsafe { &*buf.offset(idx) };
-                    let symbol_name = unsafe { std::ffi::CStr::from_ptr(swt_loc.symbol_name) };
-                    SirLoc {
-                        symbol_name: symbol_name.to_str().unwrap(),
-                        bb_idx: swt_loc.bb_idx,
-                        addr: None
-                    }
-                })
-                .collect();
-
-            unsafe { libc::free(buf as *mut libc::c_void) };
-            Ok(SirTrace::new(locs))
-        }
+        let len;
+        let buf = unsafe {
+            if !__YK_SWT_ACTIVE {
+                libc::puts("no trace recorded\n\0" as *const str as *const i8);
```

Comment:
A better error message would be something like: "attempted to stop tracing when no tracer is active"?

Reply:
By the way, why are we using libc functions here? Was is to prevent them being traced?

Reply:
> Was is to prevent them being traced?

Yes, that was the reason. It could be changed back now.

Reply:
OK. In a later PR we could convert as much of this to idiomatic rust as possible.